kernel/src/main/java/org/kframework/kompile/DefinitionStorage.java

```diff
@@ -0,0 +1,56 @@
+// Copyright (c) 2019 K Team. All Rights Reserved.
+package org.kframework.kompile;
+
+import org.kframework.parser.concrete2kore.ParseCache;
+import org.kframework.utils.BinaryLoader;
+import org.kframework.utils.file.FileUtil;
+
+import javax.inject.Inject;
+import java.io.File;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * @author Denis Bogdanas
+ * Created on 09-Nov-19.
+ */
+public class DefinitionStorage {
+    private final BinaryLoader loader;
+    private final FileUtil files;
+
+    @Inject
+    public DefinitionStorage(BinaryLoader loader, FileUtil files) {
+        this.loader = loader;
+        this.files = files;
+    }
+
+    public File getCacheFile() {
+        return files.resolveKompiled("definitionAndCache.bin");
+    }
+
+    public void save(DefinitionAndCache definitionAndCache) {
+        if (!cacheParses(definitionAndCache.compiledDefinition.kompileOptions)) {
+            definitionAndCache.parseCaches.clear();
```

Comment:
this seems suspect to me. you are clearing the cache after having run the tool, when profile rules is enabled. This means that the following run, which does not have profile rules, will not use the cache. This is probably not what you want.

Instead, I think what you should do is make it so that the loading of the cache doesn't happen if profile rules is enabled, so that it sees an empty cache even though the cache isn't really empty, but then it should save the cache as normal.

Reply:
This preserves exactly the old behavior of `KompileOptions.profileRules`. If it is enabled, parse cache is neither loaded nor saved. Besides, what you propose wouldn't change the behavior until the next `kompile`, in cases when you do `kompile --profile-rules` followed by a `kompile` without the option. I was careful here. I believe current behavior is more intuitive. You don't save what is not supposed to be used.